homeassistant/components/sense/sensor.py

```diff
@@ -23,260 +27,247 @@
-
 # Trend production/consumption variants
 TREND_SENSOR_VARIANTS = [
-    *SENSOR_VARIANTS,
-    (PRODUCTION_PCT_ID, PRODUCTION_PCT_NAME),
-    (NET_PRODUCTION_ID, NET_PRODUCTION_NAME),
-    (FROM_GRID_ID, FROM_GRID_NAME),
-    (TO_GRID_ID, TO_GRID_NAME),
-    (SOLAR_POWERED_ID, SOLAR_POWERED_NAME),
+    PRODUCTION_ID,
+    CONSUMPTION_ID,
+    NET_PRODUCTION_ID,
+    FROM_GRID_ID,
+    TO_GRID_ID,
 ]
+TREND_SENSOR_PCT_VARIANTS = [PRODUCTION_PCT_ID, SOLAR_POWERED_ID]
 
 
-async def async_setup_entry(
-    hass: HomeAssistant,
-    config_entry: SenseConfigEntry,
-    async_add_entities: AddEntitiesCallback,
-) -> None:
-    """Set up the Sense sensor."""
-    data = config_entry.runtime_data.data
-    trends_coordinator = config_entry.runtime_data.trends
-    realtime_coordinator = config_entry.runtime_data.rt
+class CoordinatorType(Enum):
+    """Types of Coordinators."""
 
-    # Request only in case it takes longer
-    # than 60s
-    await trends_coordinator.async_request_refresh()
+    TREND = "trend"
+    REALTIME = "realtime"
 
-    sense_monitor_id = data.sense_monitor_id
 
-    entities: list[SensorEntity] = []
+@dataclass(frozen=True, kw_only=True)
+class SenseSensorEntityDescription(SensorEntityDescription):
+    """Class to describe a Sensor entity."""
 
-    for device in config_entry.runtime_data.data.devices:
-        entities.append(
-            SenseDevicePowerSensor(device, sense_monitor_id, realtime_coordinator)
-        )
-        entities.extend(
-            SenseDeviceEnergySensor(device, scale, trends_coordinator, sense_monitor_id)
-            for scale in Scale
-        )
+    value_fn: Callable[[ASyncSenseable], float]
+    last_reset_fn: Callable[[ASyncSenseable], datetime | None] | None = None
+    coordinator_type: CoordinatorType
 
-    for variant_id, variant_name in SENSOR_VARIANTS:
-        entities.append(
-            SensePowerSensor(
-                data, sense_monitor_id, variant_id, variant_name, realtime_coordinator
-            )
-        )
 
-    entities.extend(
-        SenseVoltageSensor(data, i, sense_monitor_id, realtime_coordinator)
-        for i in range(len(data.active_voltage))
-    )
+@dataclass(frozen=True, kw_only=True)
+class SenseDeviceSensorEntityDescription(SensorEntityDescription):
+    """Class to describe a Sensor entity."""
 
-    for scale in Scale:
-        for variant_id, variant_name in TREND_SENSOR_VARIANTS:
-            entities.append(
-                SenseTrendsSensor(
-                    data,
-                    scale,
-                    variant_id,
-                    variant_name,
-                    trends_coordinator,
-                    sense_monitor_id,
-                )
-            )
+    value_fn: Callable[[SenseDevice], float]
+    coordinator_type: CoordinatorType
 
-    async_add_entities(entities)
 
+def value_fn_trend(scale, variant_id):
+    """Value of Trend."""
+    return lambda gateway: gateway.get_stat(scale, variant_id)
```

Comment:
I am not sure why this is a function

Reply:
It's solely to deal with typing issues with .  Having a lambda with loop vars works with:
`lambda gateway, s=scale, v=variant_id: gateway.get_stat(s, v)`
but then the function definition in the typing is wrong.  I couldn't see a clean way to do it without two lambdas or something weird like that

Reply:
How many lambdas would this give?

Reply:
It could use the partial function like this:
`value_fn=partial(lambda gateway, s, vid: gateway.get_stat(s, vid), s=scale, vid=variant_id),`
Not sure if that's clear enough, it would at least be inline

Reply:
`value_fn=lambda gateway, s, vid: gateway.get_stat(s, vid)` isn't that enough?

Reply:
No, that doesn't match the type for the description `value_fn: Callable[[ASyncSenseable], float]`

Reply:
oh right, you mean that `s` and `vid` are constants?

Reply:
Yes, they come from the loop it's in:
```
    SenseSensorEntityDescription(
        value_fn=lambda gateway, s=scale, vid=variant_id: gateway.get_stat(s, vid),
    )
    for scale in Scale
    for variant_id in TREND_SENSOR_VARIANTS
)
```
Where Scale and TREND_SENSOR_VARIANTS are constants

Reply:
Oh hmm, maybe it's worth to just throw it in a function? I do something similar in withings for the measurements